std/crypto/chacha20.zig

```diff
@@ -0,0 +1,156 @@
+           );
+}
+
+fn salsa20_wordtobyte(input: [16]u32) [64]u8 {
+    var x: [16]u32 = undefined;
+    var out: [64]u8 = undefined;
+
+    for (x) |_, i|
+        x[i] = input[i];
+    const rounds = comptime []QuarterRound{
+        Rp( 0, 4, 8,12),
+        Rp( 1, 5, 9,13),
+        Rp( 2, 6,10,14),
+        Rp( 3, 7,11,15),
+        Rp( 0, 5,10,15),
+        Rp( 1, 6,11,12),
+        Rp( 2, 7, 8,13),
+        Rp( 3, 4, 9,14),
+    };
+    comptime var j: usize = 20;
+    inline while (j > 0) : (j -=2) {
+        for (rounds) |r| {
+            x[r.a] +%= x[r.b]; x[r.d] = rotate(x[r.d] ^ x[r.a], 16);
+            x[r.c] +%= x[r.d]; x[r.b] = rotate(x[r.b] ^ x[r.c], 12);
+            x[r.a] +%= x[r.b]; x[r.d] = rotate(x[r.d] ^ x[r.a],  8);
+            x[r.c] +%= x[r.d]; x[r.b] = rotate(x[r.b] ^ x[r.c],  7);
+        }
+    }
+    for (x) |_, i|
+        x[i] +%= input[i];
+    for (x) |_, i|
+        mem.writeInt(out[4 * i .. 4 * i + 4], x[i], builtin.Endian.Little);
+    return out;
+}
+
+pub fn chaCha20(in: []const u8, key: [32]u8, nonce: [8]u8, out: *[]u8) void {
+    var ctx: [16]u32 = undefined;
+    var remaining: usize = undefined;
+    var cursor: usize = 0;
+
+    //if (in.len > out.len) {
+    //    remaining = out.len;
+    //} else
+        remaining = in.len;
+
+    comptime const c = "expand 32-byte k";
+    comptime const constant_le = []u32{
+        comptime mem.readIntLE(u32, c[0..4]),
+        comptime mem.readIntLE(u32, c[4..8]),
+        comptime mem.readIntLE(u32, c[8..12]),
+        comptime mem.readIntLE(u32, c[12..16]),
+    };
+    
+    for (constant_le) |_, i|
+        ctx[i] = constant_le[i];
+
+    ctx[4] = mem.readIntLE(u32, key[0..4]);
+    ctx[5] = mem.readIntLE(u32, key[4..8]);
+    ctx[6] = mem.readIntLE(u32, key[8..12]);
+    ctx[7] = mem.readIntLE(u32, key[12..16]);
+    ctx[8] = mem.readIntLE(u32, key[16..20]);
+    ctx[9] = mem.readIntLE(u32, key[20..24]);
+    ctx[10] = mem.readIntLE(u32, key[24..28]);
+    ctx[11] = mem.readIntLE(u32, key[28..32]);
+
+    ctx[12] = 0;
+    ctx[13] = 0;
+    ctx[14] = mem.readIntLE(u32, nonce[0..4]);
+    ctx[15] = mem.readIntLE(u32, nonce[4..8]);
+
+    while (true) {
+        var buf = salsa20_wordtobyte(ctx);
+        var count: u64 = undefined;
+
+        if (remaining < 64) {
+            var i: usize = 0;
+            while (i < remaining) : (i += 1)
+                out.*[cursor + i] = in[cursor + i] ^ buf[i];
+            return;
+        }
+
+        comptime var i: usize = 0;
+        inline while (i < 64) : (i += 1)
+            out.*[cursor + i] = in[cursor + i] ^ buf[i];
+
+        cursor += 64;
+        remaining -= 64;
+
+        count = @intCast(u64, ctx[12]) | @intCast(u64, ctx[13]) << 32;
+        count += 1;
+        ctx[12] = @intCast(u32, count & @maxValue(u32));
+        ctx[13] = @intCast(u32, count >> 32);
+    }
+}
+
+// https://tools.ietf.org/html/draft-agl-tls-chacha20poly1305-04#section-7
```

Comment:
Let's test for all available test vectors.

Reply:
the first 4 are trivial, I added the complicated one. I think we only need one of the trivial ones.

Reply:
done